sled-agent/src/services.rs

```diff
@@ -2878,211 +2927,224 @@ impl ServiceManager {
 
     // Ensures that only the following Omicron zones are running.
     //
-    // Does not record any information such that these services are
-    // re-instantiated on boot.
-    async fn ensure_all_omicron_zones<F>(
+    // This method strives to be idempotent.
+    //
+    // - Starting and stopping zones is not an atomic operation - it's possible
+    // that we cannot start a zone after a previous one has been successfully
+    // created (or destroyed) intentionally. As a result, even in error cases,
+    // it's possible that the set of `existing_zones` changes. However, this set
+    // will only change in the direction of `new_request`: zones will only be
+    // removed if they ARE NOT part of `new_request`, and zones will only be
+    // added if they ARE part of `new_request`.
+    // - Zones are not updated in-place: two zone configurations that differ
+    // in any way are treated as entirely distinct.
+    // - This method does not record any information such that these services
+    // are re-instantiated on boot.
+    async fn ensure_all_omicron_zones(
```

Comment:
I think a fair amount of complexity with the old version of this method came it trying to do a three-way merge between:

- "What was the old config (values in `old_config`)"?
- "What is the new config (values in `new_request`)"?
- "What are the set of zones that the Sled Agent knows about (`existing_zones`)"?
- (and arguably, a fourth: "What are the actual set of zones that the OS reports?")

While re-working this, I realized that the information contained within `old_config` **does not matter** -- we don't really care what the old configuration was trying to do, we really just need to move "current state of the system" to try to track "new request" as best as we can.

By cutting out both `old_config` and the "order-dependent filtering", IMO, this method is a little simpler.